### PR TITLE
Intial CLI app with ability to fetch collections

### LINKED ISSUE #13: Feature: Initial cli app with ability to print one or more collections to stdout

### PR DESCRIPTION
**Pull request by StefMa**
closes #13 

### 👇 Original description of this PR:
I started a bit with the CLI part.
Honestly, I've never used Cobra and I'm not sure if I'm doing things right 😁 
However, the following works already:
```
outline collection export [CollectionId] [FilePathToSaveTheCollectionDocuments] --apiKey --baseUrl
```

@rsjethani can you have a look at this and provide feedback? 🙂 

Questions in my head:
* Is package `main` correct? 🤔 Shouldn't it be rather `cli`? 🤔 But how would I create a binary then that executes the `main` function? 🤔 
* How do we want to define the cli? I thought we go with something like 
  * `outline` (root command)
  * `collection` (used to operatore on collection stuff)
  * `document` (in case you operator on document stuff)
  * Each of those sub commands (`collection` and `document`) could have more sub commands:
  * `collection create [name]`
  * `collection delete [id]`
  * `...`
  * `document upload [markdownFileToUpload] [collectionId]`
  * `document delete [documentId]`
  * `...`

What do you think? 🤔 
I'm not sure of such a strucutre make sense. I'm not that familiar with oultine (yet).

Thanks :v:

### CODE REVIEW

File: cli/main.go [hunk 1]
```diff
@@ -1 +1,62 @@
...
+				panic(err)
+			}
+			// TODO: Safe documents in directoryToSafe in markdown file
+			fmt.Println(directoryToSafe)
+			fmt.Printf("%+v", documents)
+		},
+	}
+
+	collectionExportCmd.Flags().StringVar(&baseUrlFlag, "baseUrl", "", "The base url to the outline instance")
+	collectionExportCmd.Flags().StringVar(&apiKeyFlag, "apiKey", "", "The outline apiKey")
```
Comment (rsjethani) [inline cli/main.go:56]: these should be associated with rootCmd

**Comment by rsjethani** [comment]
> Is package main correct? thinking Shouldn't it be rather cli? thinking But how would I create a binary then that executes the main function? thinking
Ideally the structure should be like:
- cli/main.go: bare minimum code just few line of `main()` calling something else.
- `internal/cli` package contains actual cli code for processing arguments and taking actions.

Very crude main.go can be like:
```golang
package main
import "...internal/cli"
func main(){
if err:=cli.Run();err!=nil{
  os.Exit(1)
}
}
```

To build:
```
go build   -o outcli ./cli 
```

File: cli/main.go (hunk 1 above)
Comment (StefMa) [reply cli/main.go:56]: Why? This is a sub command of the collection command. Why should it be connected to the rootCmd? 🤔

**Comment by rsjethani** [comment]
The overall subcommand structure as you explained above LGTM.

**Comment by rsjethani** [comment]
@StefMa IMO this PS should stick to the associated milestone objectives...we will export etc. in next one.

**Comment by StefMa** [comment]
Hi @rsjethani, 

please have a real look 🤓 
For me this is ready for review.
It will fix #13 

To build the cli:
```
go build -o outcli ./cli      
```

Here are the different commands:
* outline (print help)
* outline collection (print help)
* outline collection print a (crashes -> apiToken & baseUrl not provided)
* outline collection print a --apiToken [TOKEN] --baseUrl [BASE_URL] (crashes -> `a` is not valid)
* outline collection print [VALID_COLLECTION_ID... (Support for multiple) (works 🎉 )

Screenshot 2023-07-18 at 7 08 04 PM

**Review by rsjethani — CHANGES_REQUESTED** [review]
Looks good overall! Two suggestions:
1. Lets print the values in JSON format so that other tools can be chained with this cli to create more complex processing. Our primary use case of using in CI is a perfect example. Later we can introduce options asking users whether they want `pretty print` or `JSON format`.
2. Unlike Java or similar languages, single `.go` files do not create any new scope so logically related code should stay together making it easier to view the overall picture. So IMO all files under `internal/cli/` ATM should be in a single file say `cli.go`. Some reference of 100+ lines in same go file: https://github.com/google/go-github/blob/master/github/codespaces.go

File: cli/main.go [hunk 2]
```diff
@@ -1 +1,12 @@
...
+import (
+	"github.com/ioki-mobility/go-outline/internal/cli"
+	"os"
+)
+
+func main() {
+	if err := cli.Run(); err != nil {
+		os.Exit(1)
+	}
+}
```
[the remark this answers is not in the document]
Comment (StefMa) [reply cli/main.go]: https://github.com/ioki-mobility/go-outline/pull/16/commits/7ac9e4b6853556116ed9c8fce2012a92a4d28d12

File: internal/cli/cli.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+			for _, collectionId := range args {
+				collection, err := outlineClient.
+					Collections().
+					Get(outline.CollectionID(collectionId)).
+					Do(context.Background())
+				if err != nil {
+					panic(err)
+				}
+
+				collectionAsJson, err := json.Marshal(&collection)
```
Comment (rsjethani) [inline internal/cli/cli.go:75]: no need for such a big var name just calling it `b` should be enough (and actually preferred) for such a short scope IMO.

File: internal/cli/cli.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+		Run: func(cmd *cobra.Command, args []string) {
+			baseUrl, err := rootCmd.Flags().GetString(flagNameBaseUrl)
+			if err != nil {
+				panic(err)
+			}
+			apiKey, err := rootCmd.Flags().GetString(flagNameApiKey)
+			if err != nil {
+				panic(err)
+			}
+			outlineClient := outline.New(baseUrl, &http.Client{}, apiKey)
```
Comment (rsjethani) [inline internal/cli/cli.go:65]: naming the var as just `client` is enough IMO. We can see very easily its type and details very easily in a static language like go so why add extra noise :)

**Review by rsjethani — CHANGES_REQUESTED** [review]

**Review by rsjethani — CHANGES_REQUESTED** [review]
LGTM, 
However since we want to keep main branch history clean  and there are no merge commits for the same reason I would request that you add proper good looking commit message like this one https://github.com/ioki-mobility/go-outline/commit/45b59fd16542b179d101775c3959a25a35847e79 you can push an empty commit also with `git commit --allow-empty`. Lets figure out better way to do this over a call for future PRs.

**Comment by StefMa** [comment]
@rsjethani I could do a squash and merge 🤓 
What about the following as a commit message:

```
Initital CLI client to fetch collection

The cli can be invoked via the `outline` command and based on the `cobra` package.
For now, we support a `collection` subcommand to operate on anything related outline collections.
The `fetch` command on the `collection` command can be triggered to print a specific collection in json format to stdout.
```

What do you think? 🤔

**Comment by rsjethani** [comment]
@StefMa that message looks nice! Just fit it like following: message title <=50 and body >=72 chars and rebase your branch. All good then!

**Review by rsj-ioki — APPROVED** [review]
Thanks! LGTM!

**Review by rsjethani — APPROVED** [review]

**Pull request merged**